Azure.ResourceManager.Core/Resources/Location.cs

```diff
@@ -2,98 +2,445 @@
+                    CanonicalName = location;
+                    DisplayName = location;
+                    break;
+                case NameType.CanonicalName:
+                    Name = GetDefaultNameFromCanonicalName(location);
+                    CanonicalName = location;
+                    DisplayName = GetDisplayNameFromCanonicalName(location);
+                    break;
+                case NameType.DisplayName:
+                    Name = GetDefaultNameFromDisplayName(location);
+                    CanonicalName = GetCanonicalNameFromDisplayName(location);
+                    DisplayName = location;
+                    break;
+            }
         }
 
-        internal Location()
+        private enum NameType
         {
+            Name,
+            CanonicalName,
+            DisplayName,
         }
 
+        /// <summary>
+        /// Gets defaul Location object: West US.
+        /// </summary>
         public static ref readonly Location Default => ref WestUS;
 
-        public string Name { get; internal set; }
+        /// <summary>
+        /// Gets a location name consisting of only lowercase characters without white spaces or any separation character between words, e.g. "westus".
+        /// </summary>
+        public string Name { get; private set; }
 
-        public string CanonicalName { get; internal set; }
+        /// <summary>
+        /// Gets a location canonical name consisting of only lowercase chararters with a '-' between words, e.g. "west-us".
+        /// </summary>
+        public string CanonicalName { get; private set; }
 
-        public string DisplayName { get; internal set; }
+        /// <summary>
+        /// Gets a location display name consisting of titlecase words or alphanumeric characters separated by whitespaces, e.g. "West US"
+        /// </summary>
+        public string DisplayName { get; private set; }
 
         /// <summary>
+        /// Creates a new location implicitly from a string.
         /// </summary>
-        /// <param name="other"></param>
-        public static implicit operator string(Location other)
+        /// <param name="other">String to be assigned.</param>
+        public static implicit operator Location(string other)
         {
-            return other.DisplayName;
+            if (other == null)
+                return null;
+
+            var normalizedName = NormalizationUtility(other);
+            Location value;
+            if (PublicCloudLocations.TryGetValue(normalizedName, out value))
+            {
+                return value;
+            }
+
+            return new Location(other);
         }
 
-        public static implicit operator Location(string other)
+        /// <summary>
+        /// Creates a string implicitly from a Location object.
+        /// </summary>
+        /// <param name="other">Location object to be assigned.</param>
+        public static implicit operator string(Location other) => other.DisplayName;
```

Comment:
Can we point this at other.ToString() that way these are consistent in case we change the behavior

We also should need a null check here which if the unit tests pass probably means we are missing one.
if(other == null)
   return null;

Location x = null;
string y = x; //this currentlyresults in a null reference exception
Assert.IsNull(y);